core/provable-chain.go

```diff
@@ -0,0 +1,402 @@
+package core
+
+import (
+	"context"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	transfertypes "github.com/cosmos/ibc-go/v8/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
+	conntypes "github.com/cosmos/ibc-go/v8/modules/core/03-connection/types"
+	chantypes "github.com/cosmos/ibc-go/v8/modules/core/04-channel/types"
+	ibcexported "github.com/cosmos/ibc-go/v8/modules/core/exported"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// ProvableChain represents a chain that is supported by the relayer.
+//
+// It wraps primary methods of the Chain and Prover interfaces with tracing.
+// This allows the relayer to provide tracing functionality without modifying module code.
+//
+// Modules can also add custom attributes to spans. For example, a module can add attributes
+// in the GetMsgResult method as follows:
+//
+//	func (c *Chain) GetMsgResult(ctx context.Context, id core.MsgID) (core.MsgResult, error) {
+//		msgID, ok := id.(*MsgID)
+//		if !ok {
+//			return nil, fmt.Errorf("unexpected message id type: %T", id)
+//		}
+//
+//		span := trace.SpanFromContext(ctx)
+//		span.SetAttributes(core.AttributeKeyTxHash.String(msgID.TxHash))
```

Comment:
Here is an example of the `tx_hash` attribute set by `Chain.rawSendMsgs` in the tendermint module:

![image](https://github.com/user-attachments/assets/2d0de80e-6cd8-4ca1-a669-f64981c220f8)